src/main/java/com/sayweb/controller/AppController.java

```diff
@@ -39,79 +41,112 @@ public ResponseEntity<Map<String, String>> convertToWordsAndAudio(
+      e.printStackTrace();
+      response.put("error", "An unexpected error occurred.");
+      return ResponseEntity.status(500).body(response);
     }
   }
 
   @GetMapping("/audio/{fileName}")
   public ResponseEntity<InputStreamResource> getAudioFile(@PathVariable String fileName) {
     try {
-      // Use the dedicated audio directory in the container
-      File audioFile = new File(AUDIO_DIR, fileName);
+      // Ensure the directory exists
+      File audioDir = new File(AUDIO_DIR);
+      if (!audioDir.exists()) {
+        audioDir.mkdirs();
+      }
+
+      // Check and cleanup old files
+      cleanupOldFiles(audioDir);
+
+      // Generate file path
+      File audioFile = new File(audioDir, fileName);
+      if (!audioFile.exists()) {
+        // Extract number from file name if possible
+        Long number = extractNumberFromFileName(fileName);
+
+        // Convert the number to words
+        String words = SayNumber.convertToWords(number);
+
+        // Generate audiofile
+        String[] command = {"espeak-ng", "-w", audioFile.getAbsolutePath(), words};
+        Process process = Runtime.getRuntime().exec(command);
+        process.waitFor(); // Wait for the audio file to be created
+      }
+
+      // Verify the file exists after attempting to create it
       if (!audioFile.exists()) {
-        return ResponseEntity.notFound().build();
+        throw new IOException("Failed to generate the audio file.");
       }
 
       // Serve the file
       InputStreamResource resource = new InputStreamResource(new FileInputStream(audioFile));
       HttpHeaders headers = new HttpHeaders();
       headers.add("Content-Disposition", "inline; filename=" + audioFile.getName());
 
-      ResponseEntity<InputStreamResource> response =
-          ResponseEntity.ok()
-              .headers(headers)
-              .contentLength(audioFile.length())
-              .contentType(MediaType.APPLICATION_OCTET_STREAM)
-              .body(resource);
-
-      // Schedule the file for deletion after a longer delay
-      new Thread(
-              () -> {
-                try {
-                  // Wait for 5 minutes to ensure the download completes
-                  Thread.sleep(300000); // 300 seconds delay (5 minutes)
-                  audioFile.delete();
-                } catch (InterruptedException e) {
-                  e.printStackTrace();
-                }
-              })
-          .start();
-
-      return response;
-
-    } catch (IOException e) {
-      return ResponseEntity.notFound().build();
+      return ResponseEntity.ok()
+          .headers(headers)
+          .contentLength(audioFile.length())
+          .contentType(MediaType.APPLICATION_OCTET_STREAM)
+          .body(resource);
+    } catch (IOException | InterruptedException e) {
+      return ResponseEntity.status(500).body(null);
+    }
+  }
+
+  private void cleanupOldFiles(File audioDir) {
+    File[] files = audioDir.listFiles();
+    if (files != null) {
```

Comment:
Better to avoid nesting where possible. Since nested code is difficult to grasp.

```
   if (files == null) return; 

   for (File file : files) {
       if (file.isFile() && isOlderThanOneDay(file)) {
         file.delete();        
   }
 }
```